### ISSUE TITLE
identical functions not deduplicated 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#82817]
````rust
pub fn fna(s: Option<String>) -> usize {
     match s  {
         Some(s) =>{ return s.len();},
         None =>  {return 0; },
     }

}


pub fn fnb(s: Option<String>) -> usize {
     if let Some(s) = s {
         return s.len();
     } else {
         return 0;
     }
}


pub fn fnc(s: Option<String>) -> usize {
      s.map(|x| x.len()).unwrap_or(0)
}
````
https://rust.godbolt.org/z/cG3b9x

All these 3 functions do the same.
For the first two, llvm/rust realizes that they are identical and only saves one copy inside the binary, but for `fnc` this is not the case for some reason.
`fnc` is saved separately although it generates the same instructions as `fna`/`fnb`

https://rust.godbolt.org/z/Y6zazj

**Comment by klensy** [comment]
Today (rustc 1.79.0-nightly (0d8b3346a 2024-04-14)) its one fna: https://godbolt.org/z/rq9dx8n8v

**Comment by blyxyas** [comment]
Fixed currently in stable Rust 1.78. I think we can close this as succesful :)
